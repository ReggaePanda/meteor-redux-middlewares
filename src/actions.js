import { createAction } from './utils';

export const REGISTER_REACTIVE_SOURCE = 'REGISTER_REACTIVE_SOURCE';
export const registerReactiveSource = createAction(REGISTER_REACTIVE_SOURCE);

export const STOP_SUBSCRIPTION = 'STOP_SUBSCRIPTION';
export const stopSubscription = createAction(STOP_SUBSCRIPTION);

export const START_SUBSCRIPTION = 'START_SUBSCRIPTION';
export const startSubscription = createAction(START_SUBSCRIPTION);
