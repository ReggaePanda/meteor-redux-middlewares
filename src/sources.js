import { REGISTER_REACTIVE_SOURCE } from './actions';
import { actionCase, hasGet, hasSubscribe, hasKey, errorWith } from './utils';
import runContextual from './runContextual';

const computations = {};

export default tracker => ({ dispatch }) => next => action => {
  const throwIfNot = errorWith(action);

  if (action.type === REGISTER_REACTIVE_SOURCE) {
    const run = () => {
      throwIfNot(hasKey,
        'A registerReactiveSource action needs a `key` string to identify tracked source'
      );

      throwIfNot(hasGet,
        'A registerReactiveSource action needs a `get` function to load data'
      );

      throwIfNot(x => !hasSubscribe(x),
        'Use a startSubscription action to start a subscription'
      );

      const { key } = action.payload;

      if (computations[key]) {
        computations[key].stop();
      }

      computations[key] = tracker.autorun(() => {
        dispatch({
          type: `${actionCase(key)}_REACTIVE_SOURCE_CHANGED`,
          payload: action.payload.get(),
        });
      });
    };

    runContextual(run);
  }

  return next(action);
};
