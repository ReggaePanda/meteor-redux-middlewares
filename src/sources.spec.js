import configureMockStore from 'redux-mock-store';
import sources from './sources';
import { registerReactiveSource } from './actions';


/**
 * See src/__mocks__/mocks.js
 *
 * The results array imported here is used to hold tags pushed from various
 * functions to ensure the sequence is as expected
 * (the tags are in pirate speak)
 *
 */
const {
  get,
  results,
  tracker,
} = require('mocks')();

const register = registerReactiveSource({ key: 'mates', get });

const store = configureMockStore([sources(tracker)])({});
store.dispatch(register);
store.dispatch(register);

it('should handle a registerReactiveSource action', () => {
  expect(store.getActions()).toMatchSnapshot();
});

it('should fire in the correct sequence', () => {
  expect(results).toMatchSnapshot();
});
