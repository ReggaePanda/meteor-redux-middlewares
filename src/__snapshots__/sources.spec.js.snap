exports[`test should fire in the correct sequence 1`] = `
Array [
  "i got me data!",
  "my computation be stopped now!",
  "i got me data!",
]
`;

exports[`test should handle a registerReactiveSource action 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "id": "treasure",
      },
    ],
    "type": "MATES_REACTIVE_SOURCE_CHANGED",
  },
  Object {
    "meta": Object {},
    "payload": Object {
      "get": [Function],
      "key": "mates",
    },
    "type": "REGISTER_REACTIVE_SOURCE",
  },
  Object {
    "payload": Array [
      Object {
        "id": "treasure",
      },
    ],
    "type": "MATES_REACTIVE_SOURCE_CHANGED",
  },
  Object {
    "meta": Object {},
    "payload": Object {
      "get": [Function],
      "key": "mates",
    },
    "type": "REGISTER_REACTIVE_SOURCE",
  },
]
`;
